# The Common Fund Data Ecosystem's Crosscut Metadata Model (CFDE C2M2)

This document introduces the Crosscut Metadata Model
([C2M2](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#c2m2)), a
flexible metadata standard for describing experimental resources in
biomedicine and related fields. The Common Fund Data Ecosystem
group is creating a new computing infrastructure, with C2M2 as its
organizing principle, to offer the health research community an
unprecedented array of intersectional data tools. The C2M2 system will
connect researchers with scale-powered statistical analysis
methods; deep, seamless searching across experimental data generated by
different projects and organizations; and new ways to aggregate
and integrate experimental data from different sources to
facilitate scientific replication and to drive new discoveries.

Using this new infrastructure, Common Fund data coordinating centers
([DCCs](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#dcc)) will
share structured, detailed information ([metadata](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#metadata))
about their experimental resources with the research
community at large, widening and deepening access to usable
observational data. One immediate consequence will be a
drastic simplification in the effort required to perform
meta-analysis of results from multiple independent teams
studying similar health-related phenomena.

## DCC Metadata Submissions

DCCs collect and provide metadata submissions
(C2M2 [instances](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#c2m2-instance))
to CFDE describing experimental resources within their purview. Each
submission is a set of tab-separated value files
([TSVs](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#tsv));
precise formatting requirements for these filesets are specified by
JSON Schema documents, each of which is an instance of the
[Data Package](http://frictionlessdata.io/docs/data-package/)
meta-specification published by the
[Frictionless Data](http://frictionlessdata.io/)
group. The Data Package meta-specification is a toolkit for
defining format and content requirements for files so that
automatic validation can be performed on those files, just
as a database management system stores definitions for
database tables and automatically validates incoming data
based on those definitions. Using this toolkit, the C2M2 JSON
Schema specifications lay out
[foreign-key](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#foreign-key)
relationships between metadata fields (TSV columns), rules
governing missing data, required content types for particular fields,
and other similar database management constraints to define
basic structural integrity for C2M2 metadata submissions.
During the
[C2M2 ingestion process](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#c2m2-ingestion-process),
the C2M2 software infrastructure uses these specifications
to automatically validate format compliance and submission integrity,
prior to loading metadata into its central database. Once loaded,
metadata are used to fuel downstream services like search results,
customized statistical summaries, dynamic display graphics, and asset
browsing within experimental resource collections.

## C2M2 Levels

CFDE offers DCCs three alternative metadata submission
formats (C2M2 Levels 0, 1 and 2), each of which is automatically
interoperable with the entire C2M2 software ecosystem. Levels
are tiered primarily according to increasing complexity.
The general idea is that DCC resource collections can be
represented quickly (and thus begin driving downstream
applications quickly) using metadata encoded at lower (simpler)
C2M2 levels: over time, and as feasible, DCC data managers
can upgrade their C2M2 metadata submissions by expanding
into higher levels.

C2M2 Levels 0, 1 and 2 are increasingly large and complex variants of the
same metadata model (C2M2): each level is a defined collection of
data tables (encoded as TSVs: see above on
[submissions](#dcc-metadata-submissions)). Level 0 defines just
one short `file` table; Level 1 provides a larger `file` table than
Level 0 (more fields), adds more tables describing other
basic biomedical resource concepts including `biosample`,
`subject`, and `project`, and introduces ways to express simple
relationships between records in different tables; Level 2 is
an ever-growing mature metadata interchange standard, customized
to advanced DCC metadata needs: it contains all Level 1 information,
plus a library of more detailed metadata objects and extensions to
existing objects. Lower levels are strict subsets of higher ones:
Level 1's `file` table contains all Level 0 `file` fields and
more; Level 2's `biosample` is an expanded superset of the Level
1 `biosample`, etc. Upgrades from one level to the next are
therefore limited by design to be done only by expansion: moving
up a level will not require DCCs to make changes to any metadata
already provided.

A foundational purpose of the C2M2 system is to facilitate
metadata harmonization: finding ways wherever possible to
represent comparable things in standard ways, without compromising
meaning, context or accuracy. In addition to complexity
management, C2M2 levels are also intended to roughly encapsulate
groups of concepts according to increasing harmonization
difficulty.

Some examples, sorted by increasingly heavy challenges to harmonization:

* All DCCs have file resources, describable (at a _very_ high level)
in a standard, noncontroversial way (size + filename: [Level 0](#level-0)).
* In addition to data files, virtually all DCCs deal in some way
with biosamples and/or subjects: metadata describing basic aspects
of these common concepts can be fairly (if still quite broadly) expressed by a
shared model ([Level 1](#level-1)), thereby enabling the beginnings of cross-dataset search
and analysis.
* Many DCCs have at least some specialized data, unique to their
own spheres of operation, which (at least for some time) will not be
meaningful candidates for system-level harmonization ([Level 2](#level-2) includes specialized
extension modeling).

Most DCCs already have some form of internal metadata model in use
for their own curation operations. C2M2 integration of similar but
distinct packages of important information, taken from multiple
independently-developed custom DCC metadata systems (including e.g.
metadata describing people and organizations, data provenance relationships,
experimental protocols, protected data, or detailed event
sequences), will require ongoing, iterative, case-based design
and consensus-driven decision-making, often coordinated across multiple
independent research groups. Design and decision-making in such
contexts will require long-term planning, testing
and execution. Metadata difficult (or even impossible) to
integrate and harmonize is thus handled as part of the ongoing
evolution and expansion of Level 2, leaving Level 1 tasked with
supporting relatively universal, simple and uncontroversial
metadata concepts to maintain streamlined development and deployment of
important core metadata packages without unnecessarily blocking feasible
tasks to wait on more expensive custom integration.

With the design of C2M2, we are splitting the difference
between the ease of evolution inherent in a simple model and
the operational power provided to downstream applications by more
complicated and difficult-to-maintain extended frameworks.

Modeling and data wrangling are always difficult, even for
experts. Part of the goal of the level system
is to compartmentalize the C2M2 model so as to maintain flexibility --
especially during developmental phases -- in order to best
accommodate mutual learning between DCCs and CFDE as the
construction of this federated metadata system progresses. It
is generally far more expensive and error-prone to repeatedly
change a complex, over-built, inseparable, monolithic model than it is
to build one gradually from a simpler core of agreement which
can be relatively quickly stabilized while more specialized
branches are built in parallel and transitioned into more general use.

At any given moment, participating Common Fund DCCs will
span a broad range of experience and available funding,
based on mission details and lifecycle phases. DCCs with
advanced, operationalized metadata modeling systems of
their own should not encounter arbitrary barriers to
C2M2 support for more extensive relational modeling of
their metadata if they want it; CFDE will maintain such
support by iteratively refining Level 2
according to needs identified while working with DCCs
already wielding complex metadata models. Newer or smaller
DCCs, by contrast, may not have enough readily-available
information to feasibly describe their experimental resources
using Level 2 structures (either existing or proposed): C2M2
Level 1 thus also aims to actively support such cases by
offering simpler but still well-structured metadata models
where metadata has already harmonized across other DCCs, lowering
barriers to rapid entry into the data ecosystem and meaningful
participation in downstream services.

A C2M2 topic requiring special attention is the use of _identifiers_.

--------------------------------------------------------------------------------

### C2M2 identifiers

* _Two complementary identifier slots for DCC-issued records_
   * `persistent_id`: _persistent and resolvable ID (ideal, but optional)_
   * `id_namespace`, `local_id`: _2-part key is as least-common denominator_
* _The optional persistent identifier is a DOI, ARK, MINID, etc._
* _The 2-element composite identifier conveys fragments of a record URI_
   * `local_id` _bears a name from some namespace, e.g. an accession ID_
   * `id_namespace` _specifies which namespace, i.e. left-hand side of a URI_
   * _concatenation of namespace + local yields the full record URI_
   * _URIs are cheap and easy: low barrier to entry, no hosting requirements_
* _To steer users directly to data, we require_ `persistent_id`_!_
* _Otherwise, we can steer users towards the DCC contact._
* _For consistency, we repeat this scheme on several entity tables._

_Not all DCCs arrive having adopted robust, persistent and resolvable identifiers,
or we could just mandate their use to identify all records. We need to
support different DCC identifier maturity levels, and so we need to keep_
`persistent_id` _optional._

_We add a 2-part composite identifier as something which can lower the barrier
to entry: this is a URI that has been split into two parts. The web rules for
forming URIs are very flexible: it takes essentially no cost and no real
infrastructure to generate URIs, which are just rules for namespace hygiene
laying out how a DCC can claim a namespace out of the ether, ideally rooted in
some other basic name they own such as a DNS name (website domain) or even
just an email address._

_(For those who don’t know, the difference between a URI and a URL is that a
URI is just an identifier.  It doesn’t have to address any actual web server
which responds to messages, so you can produce URIs all day long and never
worry about hosting costs, availability, etc.)_

_The core idea is that most DCCs will have some accession ID or other
locally-unique key for their assets already. They can copy that value
into the_ `local_id` _part or define some other mapping of their own
choosing, there. Then, they can either define a new_ `id_namespace` _or pass
through an applicable, existing URI prefix as the_ `id_namespace` _component.
As long as each DCC follows the web rules to use an_ `id_namespace` _they “own”,
collisions are automatically avoided (where two DCCs might try to use the
same composite identifier)._

_We also allow for the use of persistent, resolvable identifiers for things
which are not really data. We suspect that many of the same practical
benefits of these identifiers for data might apply to other entity
types, even if resolution might lead to landing pages and contact
info which at most guides users to documentation or other possible
actions they might take in the real world or in the parallel
bureaucratic world. (Attempting to directly download subjects or
biosamples might trigger some unexpected and unpleasant -- although
possibly comical -- downstream side effects.)_

--------------------------------------------------------------------------------

* _For a DCC already issuing persistent, resolvable IDs_
   * `persistent_id`: _fill with canonical ID_
   * `id_namespace`, `local_id`: _fill with (split) copy of canonical ID_
* _For a DCC already issuing relatively stable URIs_
   * `persistent_id`: _leave blank until ready_
   * `id_namespace`, `local_id`: _fill with (split) copy of URI_
* _For a DCC already issuing local accession IDs_
   * `persistent_id`: _leave blank until ready_
   * `local_id`: _fill with local accession ID_
   * `id_namespace`: _choose an appropriate namespace URI-prefix_
* _For a DCC without local ID stability_
   * _Need to invent something approximating accession ID and proceed as above_

_If a DCC already uses persistent identifiers such as DOIs, ARKs, or_
_other short identifiers resolvable by some name-to-thing service then they can_
_just put that value into all the identifier fields:_

* _verbatim in the_ `persistent_id` _field so that consumers know a_ `persistent_id` _is available for this record_
* _split and replicated into the_ `id_namespace` _and_ `local_id` _fields so that the core C2M2 requirement for a record key is met_
* _there is no need for the DCC to issue and juggle two separate identifier formats, but the compromise C2M2 format requires the fields to be populated_

_If a DCC already uses URIs or URLs for entities having a one-to-one
correspondence to C2M2 record concepts, they can split that URI or
URL into a namespace prefix part and a final local part and use
those values to fill the 2-part composite record identifier. They would
probably leave_ `persistent_id` _blank._

_If a DCC only has local identifiers for such entities, they can put
that in the_ `local_id` _part and then fabricate a new_ `id_namespace`
_representing their DCC or the sub-organizational scope where these
local identifiers are indeed unique. The same _ `id_namespace` _should
be reused for all peer records with local parts coming from the
same DCC naming system._

_Having stable identifier management is, unavoidably, basic
“table stakes” for a DCC to produce reusable data. Archiving
and indexing options like those provided by C2M2 are extremely
limited in the absence of some method of persistent
bookkeeping to track inventory: persistent identifiers are
a core requirement precisely because without them, federated
views of resource metadata from multiple DCCs cannot be maintained._

--------------------------------------------------------------------------------

### Level 0

C2M2 Level 0 defines a **minimal valid
[C2M2 instance](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#c2m2-instance).**
Data submissions at this level of
[metadata richness](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#c2m2-richness-levels)
will be the easiest to produce, and will support the simplest available
functionality implemented by downstream applications.

#### Level 0 submission process: overview

Metadata submissions at Level 0 will consist of a single
[TSV file](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#tsv)
describing a **collection of digital
[files](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#file)**
owned or managed by a
[DCC](https://cfde-published-documentation.readthedocs-hosted.com/en/latest/CFDE-glossary/#dcc).
The properties listed for the Level 0 `file` entity (see below for
diagram and definitions) will serve as the TSV's column headers; each TSV
row will represent a single file. The Level 0 TSV itself thus represents a
**manifest or inventory** of digital files that a DCC wants to introduce
into the C2M2 metadata ecosystem.

This level encodes the most basic file metadata: its use by downstream applications will be
limited to informing the least specific level of data accounting, querying and reporting.

|_Level 0 model diagram_|
|:---:|
|![Level 0 model diagram](../draft-C2M2_ER_diagrams/Level-0-C2M2-model.png "Level 0 model diagram")|

#### Level 0 technical specification: properties of the `file` entity

**Required: `id_namespace` `local_id` `sha256|md5`**

|property|description|
|:---:|:---|
| `id_namespace` | String **identifier devised by the DCC managing this `file`** (cleared by CFDE-CC to avoid clashes with any preexisting `id_namespace` values). The value of this property will be used together with `local_id` as a **composite key structure formally identifying Level 0 `file` entities** within the total C2M2 data space. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `local_id` | Unrestricted-format **string identifying this `file`**: can be any string as long as it **uniquely identifies each `file`** within the scope defined by the accompanying `id_namespace` value. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `persistent_id` | **A permanent, resolvable URI permanently attached to this `file`**, meant to serve as a permanent address to which landing pages (which summarize metadata associated with this `file`) and other relevant annotations and functions can optionally be attached, including information enabling resolution to a network location from which the `file` can be downloaded. **Actual network locations must not be embedded directly within this identifier**: one level of indirection is required in order to protect `persistent_id` values from changes in network location over time as files are moved around. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `size_in_bytes` | The **size of this `file` in bytes**. This varies (even for "copies" of the same `file`) across differences in storage hardware and operating system. CFDE does not require any particular method of byte computation: precise, reproducible file size integrity metadata will be provided in the form of checksum data in the `sha256` and/or `md5` properties. `size_in_bytes` will instead underpin automatic reporting of approximate storage statistics across different C2M2 collections of DCC metadata. |
| `sha256` | **CFDE-preferred** file checksum string: the output of the SHA-256 cryptographic hash function after being run on this `file`. One or both of `sha256` and `md5` is required. |
| `md5` | **Permitted** file checksum string: the output of the MD5 message-digest algorithm after being run as a cryptographic hash function on this `file`. One or both of `sha256` and `md5` is required. (CFDE recommends SHA-256 if feasible, but we recognize the nontrivial overhead involved in recomputing these hash values for large collections of files, so if MD5 values have already been generated, CFDE will accept them.) |
| `filename` | A filename with no prepended PATH information. |

#### Level 0 technical specification: JSON Schema and example TSVs

A JSON Schema document (implementing
[Frictionless Data](https://frictionlessdata.io/)'s
"[Data Package](https://frictionlessdata.io/data-package/)"
container meta-specification) defining the Level 0 TSV can be found
[here](../draft-C2M2_JSON_Schema_datapackage_specs/C2M2_Level_0.datapackage.json);
an example Level-0-compliant TSV submission collection can be found
[here](../draft-C2M2_example_submission_data/HMP__sample_C2M2_Level_0_bdbag.contents/file.tsv)
(just the `file.tsv` portion) and
[here](../draft-C2M2_example_submission_data/HMP__sample_C2M2_Level_0_bdbag.tgz)
(as a packaged BDBag archive).

--------------------------------------------------------------------------------

### Level 1

C2M2 Level 1 models **basic experimental resources and associations between them**.
This level of metadata richness is more difficult to produce than Level 0's flat
inventory of digital file assets. As a result, Level 1 metadata offers users more powerful
downstream tools than are available for Level 0, including
   * faceted searches on a (small) set of biologically relevant features (like anatomy
   and taxonomy) of experimental resources like `biosample` and `subject`
   * organization of summary displays using subdivisions of experimental metadata
   collections by `project` (grant or contract) and `collection` (any scientifically
   relevant grouping of resources)
   * basic reporting on changes in metadata over time, tracking (for example)
   creation times for `file` and `biosample`

C2M2 Level 1 is designed to offer an intermediate tier of difficulty, in terms of
preparing metadata submissions, between Level 0's basic digital inventory
and the full intricacy of Level 2. Accordingly, we have reserved several
modeling concepts -- requiring the most effort to produce and maintain -- for
Level 2. The following are **not modeled at Level 1**:

   * any and all **protected data**
   * documentation of  **experimental protocols**
   * event-based resource generation/**provenance networks**
   * detailed information on **organizations and people** governing the research
   being documented
   * a **comprehensive suite** of options to model **scientific attributes of
   experimental resources**
      * full collection of features like anatomy, taxonomy, and assay type,
      plus formal vocabularies to describe them
      * prerequisite to offering research users deep and detailed search possibilities

#### Level 1 submission process: overview

_Build the core C2M2 entity tables (black) and the C2M2 container tables (blue)
shown in the diagram below, and fill out the DCC contact sheet (grey).
Once you've built the core entity tables, the green tables can be built
automatically using our
[term-scanner script](../draft-C2M2_external_CV_term_table_generator_script/build_term_tables.py),
which will collect all relevant CV terms used throughout your core entity tables and
will create the corresponding green term-index tables, using data loaded from
versioned, whole-CV reference documents (like OBO files)._

_In the case of any unpopulated tables (no_ `collection` _records, for example, are
required for model compliance), please create the relevant TSV files anyway,
with just one tab-separated header line containing the empty table's column
names. (In contrast to simply omitting the blank table file, the recommended practice
instead explicitly distinguishes the case in which no data is being submitted
for a given table from the case in which a table has been omitted by mistake.)_

_Color key:_

* _Black: core entities and the direct associative relationships between them (plus_
`subject` _<->_ `subject_role_taxonomy` _)_
* _Blue: containers (projects and collections) and their containment relationships_
* _Green: external CVs (term & display-decoration tracking tables) and their foreign-key relationships_

|_Level 1 model diagram_|
|:---:|
|![Level 1 model diagram](../draft-C2M2_ER_diagrams/Level-1-C2M2-model.png "Level 1 model diagram")|

#### Level 1 technical specification

##### Core entities

   * **`file` revisited** _(superset additions: cf. below, §"Common entity fields" and also §"Controlled
   vocabularies and term tables")_
   * **`biosample` introduced** _(also cf. below, §"Common entity fields" and §"Controlled vocabularies and
   term tables")_
      * _Level 1 models_ `biosample`_s as abstract materials that are directly consumed
      by one or more analytic processes. Simple provenance relationships -- between each
      such_ `biosample` _and the_ `subject` _from which it was originally derived, as well
      as between each_ `biosample` _and any_ `file`_s analytically derived from it -- are
      represented using association tables, with one such table dedicated to each
      relationship type (cf. below, §"Association tables: inter-entity linkages").
      Actual DCC-managed provenance metadata will sometimes (maybe always) represent more complex and
      detailed provenance networks: in such situations, chains of "_`this` _produced_
      `that`_" relationships too complex to model at Level 1 will need to be
      transitively collapsed. As an example: let's say a research team collects a
      cheek-swab sample from a hospital patient; subjects that swab sample to several
      successive preparatory treatments like centrifugation, chemical ribosomal-RNA
      depletion and targeted amplification; then runs the final fully-processed
      remnant material through a sequencing machine, generating a FASTQ sequence
      file as the output of the sequencing process. In physical terms our team
      will have created a series of distinct material samples, connected one to another
      by (directed) "_`X` `derived_from` `Y`_" relationships, represented as a (possibly
      branching) graph path (in fully general terms, a directed acyclic graph) running
      from a starting node set (here, our original cheek-swab sample) through intermediate
   	nodes (one for each coherent material product of each individual preparatory process)
   	to some terminal node set (in our case, the final-stage, immediately-pre-sequencer
   	library preparation material). C2M2 Level 2 will offer metadata structures to model
   	this entire process in full detail, including representational support for all
   	intermediate_ `biosample`_s, and for the various preparatory processes involved.
   	For the purposes envisioned to be served by Level 1 C2M2 metadata, on the other hand,
   	only_ `subject` _<->_ `some_monolothic_stuff` _<->_ `(FASTQ) file` _can and should be
   	explicitly represented._
         * _The simplifications here are partially necessitated by the fact that
   	   event modeling has been deliberately deferred to C2M2 Level 2: as a result,
   	   the notion of a well-defined "chain of provenance" is not modeled at
   	   this C2M2 Level. (More concretely: Level 1 does not represent
   	   inter-_`biosample` _relationships.)_
         * _The modeling of details describing experimental processes has also been
         assigned to Level 2._
         * _With both of these (more complex) aspects of experimental metadata
         masked at C2M2 Level 1, the most appropriate granularity at which a Level 1_
         `biosample` _entity should be modeled is as an abstract "material phase"
         (possibly masking what is in reality a chain of multiple distinct materials)
         that enables an analytic (or observational or other scientific) process (which
         originates at a_ `subject` _) to move forward and ultimately produce one or
         more_ `file`_s._  
      * _In practice, a Level 1 C2M2 instance builder facing such a situation
   	might reasonably create one record for the originating_ `subject` _; create one_
   	`biosample` _entity record; create a_ `file` _record for the FASTQ file produced
   	by the sequencing process; and hook up_ `subject` _<->_ `biosample` _and_
   	`biosample` _<->_ `file` _relationships via the corresponding association tables
   	(cf. below, §"Association tables: inter-entity linkages")._
         * _In terms of deciding (in a well-defined way) specifically which native DCC
         metadata should be attached to this Level 1_ `biosample` _record, one
         might for example choose to import metadata (IDs, etc.) describing the
         final pre-sequencer material. The creation of specific rules governing maps
         from native DCC data to (simplified, abstracted) Level 1 entity records
         is of necessity left up to the best judgment of the serialization staff
         creating each DCC's Level 1 C2M2 ETL instance; we recommend consistency,
         but beyond that, custom solutions will have to be developed to handle
         different data sources. Real-life examples of solution configurations
	 will be published (as they are collected) to help inform decisionmaking,
	 and CFDE staff will be available as needed to help create mappings between
	 the native details of DCC sample metadata and the approximation that is
	 the C2M2 Level 1_ `biosample` _entity._
         * _Note in particular that this example doesn't preclude attaching multiple_
         `biosample`_s to a single originating_ `subject`_; nor does it preclude modeling a
         single_ `biosample` _that produces multiple_ `file`_s._
         * _Note also that the actual end-stage material prior to the production of a_
         `file` _might not always prove to be the most appropriate metadata source from
         which to populate a corresponding_ `biosample` _entity. Let's say a
         pre-sequencing library prepration material_ `M` _is divided in two to
         produce derivative materials_ `M1` _and_ `M2` _, with_ `M1` _and_ `M2` _then
         amplified separately and sequenced under separate conditions producing_
         `file`_s_ `M1.fastq` _and_ `M2.fastq` _. In such a case -- depending on
         experimental context -- the final separation and amplification processes
         producing_ `M1` _and_ `M2` _might reasonably be ignored for the purposes
         of Level 1 modeling, instead attaching a single (slightly upstream)_
         `biosample` _entity -- based on metadata describing_ `M` _-- to both_ `M1.fastq`
         _and_ `M2.fastq`_. As above, final decisions regarding detailed rules
         mapping native DCC data to Level 1 entities are necessarily left to
         DCC-associated investigators and serialization engineers; CFDE staff will be available as needed to offer
         feedback and guidance when navigating mapping issues._
   * **`subject` introduced** _(also cf. below, §"Common entity fields" and §"Taxonomy and the `subject` entity")_
      * _The Level 1_ `subject` _entity is a generic container meant to represent any biological
      entity from which a Level 1_ `biosample` _can be generated (the notion of_ `biosample`_s
      being generated by other_ `biosample`_s is more appropriately modeled at C2M2
      Level 2: cf. §"_`biosample` **introduced**_", immediately above)_
      * _Alongside shared metadata fields (cf. below, §"Common entity fields") and inter-entity
      associations (cf. below, §"Association tables: inter-entity linkages"), C2M2
      Level 1 models two additional details specific to_ `subject` _entities:_
         * _internal structural configuration (referred to as_ `subject_granularity` _and
         included in each_ `subject` _record as one of an enumerated list
	 of categorical value codes (for concepts like, e.g., "single organism,"
	 "microbiome," "cell line") -- reference list of granularity terms (with descriptions) is given
	 [here](../draft-C2M2_internal_CFDE_CV_tables/subject_granularity.tsv)_
         * _taxonomic assignments attached to subcomponents ("roles," another ontological enumeration
	 listed [here](../draft-C2M2_internal_CFDE_CV_tables/subject_role.tsv) for reference) of_ `subject` _entities, e.g. "cell line ancestor ->
         NCBI:txid9606" or "host (of host-pathogen symbiont system) -> NCBI:txid10090":
         this is accomplished via the_ `subject_role_taxonomy` _categorical association table
         (cf. below, §"Association table: taxonomy and the_ `subject` _entity: the_ `subject_role_taxonomy` _table")_
      * _all other_ `subject`_-specific metadata -- including any protected data -- is deferred by
      design to Level 2_

##### Common entity fields

The following properties all have the same meaning and function across
the various entities they describe (`file`, `biosample`, `project`, etc.).

|property|description|
|:---:|:---|
| `id_namespace` | String **identifier devised by the DCC managing this entity** (cleared by CFDE-CC to avoid clashes with any preexisting `id_namespace` values). The value of this property will be used together with `local_id` as a **composite key structure formally identifying Level 1 C2M2 entities** within the total C2M2 data space. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `local_id` | Unrestricted-format **string identifying this entity**: can be any string as long as it **uniquely identifies each entity** within the scope defined by the accompanying `id_namespace` value. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `persistent_id` | **A permanent, resolvable URI permanently attached to this entity**, meant to serve as a permanent address to which landing pages (which summarize metadata associated with this entity) and other relevant annotations and functions can optionally be attached, including information enabling resolution to a network location from which the entity can be viewed, downloaded, or otherwise directly investigated. **Actual network locations must not be embedded directly within this identifier**: one level of indirection is required in order to protect `persistent_id` values from changes in network location over time as entity data is moved around. (See [C2M2 identifiers](#c2m2-identifiers) for discussion and examples.) |
| `creation_time` | An ISO 8601 / RFC 3339 (subset)-compliant timestamp documenting this entity's creation time (or, in the case of a `subject` entity, the time at which the `subject` was first documented by the `project` under which the `subject` was first observed): **`YYYY-MM-DDTHH:MM:SS±NN:NN`**, where<br><ul><li>**`YYYY`** is a four-digit Gregorian **year**</li><li>**`MM`** is a zero-padded, one-based, two-digit **month** between `01` and `12`, inclusive</li><li>**`DD`** is a zero-padded, one-based, two-digit **day** of the month between `01` and `31`, inclusive</li><li>**`HH`** is a zero-padded, zero-based, two-digit **hour** label between `00` and `23`, inclusive (12-hour time encoding is specifically prohibited)</li><li>**`MM`** and **`SS`** represent zero-padded, zero-based integers between `00` and `59`, inclusive, denoting Babylonian-sexagesimal **minutes** and **seconds**, respectively</li><li>**`±`** denotes exactly one of `+` or `-`, indicating the direction of the offset from GMT (Zulu) to the local time zone (or `-` in the special case encoded as `-00:00`, in which the local time zone is unknown or not asserted)</li><li>**`NN:NN`** represents the **hours:minutes** differential between GMT/Zulu and the local time zone context of this `creation_time` (qualified by the preceding `+` or `-` to indicate offset direction), with `-00:00` encoding the special case in which time zone is unknown or not asserted (`+00:00`, by contrast, denotes the GMT/UTC/Zulu time zone itself)</li></ul><br>Apart from the **time zone** segment of `creation_time` (**`±NN:NN`**, just described) and the **year** (**`YYYY`**) segment, **all other constituent segments of `creation_time` named here may be rendered as `00` to indicate a lack of available data** at the corresponding precision.<ul><li>_We are aware (and unconcerned) that this technically renders one particular_ **`HH:MM:SS`** _value --_ "`00:00:00`" _-- ambiguous. Forestalling this ambiguity (by allowing select omissions of constituent sub-segments of_ `creation_time` _string values as an alternative mechanism to denote missing data, or by introducing nonstandard and increasingly artificial special-case encodings like_ "`99:99:99`"_) was determined to be of less immediate concern than maintaining the technical advantages conferred by the (stronger) constraint of requiring a fixed-length_ `creation_time` _string that remains fully conformant with (a constrained subset of) the RFC 3339 standard. The canonical C2M2 interpretation of_ "`00:00:00`" _is thus explicitly defined to be "_**`HH:MM:SS`** _information unknown" and not "exactly midnight."_</li></ul> |
| `abbreviation`, `name` and `description` | _Values which will be used, unmodified, for contextual display throughout portal and dashboard user interfaces: severely restricted, whitespace-free_ `abbreviation` _(must match_ `/[a-zA-Z0-9_]*/`_); terse but flexible_ `name` _; abstract-length_ `description` |

##### Containers

C2M2 Level 1 offers two ways -- `project` and `collection` -- to denote groups of
related metadata entity records representing core (`file`/`subject`/`biosample`)
experimental resources.

   * `project`
      * _unambiguous, unique, named, most-proximate research/administrative
      sphere of operations that **first generates** each experimental resource
      (_`file`/`subject`/`biosample`_) record_
      * _conceptually rooted in -- but not necessarily mapped one-to-one from
      -- a corresponding hierarchy of grants, contracts or other **important
      administrative subdivisions of primary research funding**_
      * `project` _attribution is **required** for core resource entity types: use
      FK specified in_ `file`/`biosample`/`subject` _entity records to encode these attributions_
      * `project`s _**can be nested** (via the_ `project_in_project` _association
      table: cf. below, §"Association tables: expressing containment relationships")
      into a hierarchical (directed, acyclic) network, but one and only one_
      `project` _node in one and only one_ `project` _hierarchy can be attached
      to each core entity record._
      * _by convention, for C2M2 Level 1, one artificial_ `project` _node must
      be created and identified as the root (topmost ancestor) node of each
      DCC's_ `project` _hierarchy: this node will represent the DCC itself: it
      is referenced directly (via foreign key) by the_ `primary_dcc_contact` _table,
      and serves as an anchor point for creating roll-up summaries or other
      aggregations of C2M2 metadata arranged according to managing DCC._
   * `collection`
      * _**contextually unconstrained:** a generalization of the "dataset" concept
      which additionally and explicitly supports the inclusion of elements
      (C2M2 metadata entities) representing_ `subject`_s and_ `biosample`_s_
      * _**wholly optional**: Level 1 C2M2 serialization of DCC metadata need not
      necessarily include any_ `collection` _records or attributions_
      * _**membership** of C2M2 entities in_ `collection`_s is encoded using the
      relevant association tables (cf. below, §"Association tables: expressing containment relationships")_
      * _used to describe the **federation of any set of core resource entities (and,
      recursively, other**_ **`collection`**_**s)** across inter-_`project` _boundaries
      (or across inter-DCC boundaries, or across any other structural
      boundaries used to delimit or partition areas of primary purview or
      provenance, or crossing no such boundaries at all)_
      * _unconstrained with respect to "defining entity"_
         * _**may optionally** be attributed to a (defining/generating) C2M2_ `project` _record_
            * _this attribution is **optional and (when null) will not always even
            be well-defined**: the power to define new_ `collection`_s,
            on an ongoing basis, will be offered to all (approved? registered?) members
            of the interested research community at large, without being specifically
            restricted to researchers or groups already operating under the
            auspices of a well-defined_ `project` _entity in the C2M2 system._
         * _this configuration is meant to facilitate data/metadata reuse and
         reanalysis, as well as to provide a specific and consistent anchoring
         structure through which authors anywhere can create (and study and cite)
         newly-defined groupings of C2M2 resources, independently of their original
         provenance associations. (FAIRness is generally increased by
	 provisioning for consistent reference frameworks.)_

##### Association tables: expressing containment relationships

   * `project_in_project`
   * `collection_in_collection`
   * `file_in_collection`
   * `subject_in_collection`
   * `biosample_in_collection`
   
   _These tables are used to express basic containment relationships like "this_ `file` _is in
   this_ `collection`_" or "this_ `project` _is a sub-project of this other_
   `project`_." The record format for all of these tables specifies four fields:_
   
   * _two (an_ `id_namespace` _and a_ `local_id`_) encoding a foreign key representing
   the **containing**_ **`project`** _**or**_ **`collection`**_, and_
   * _two (another_ {`id_namespace`, `local_id`} _pair) acting as a foreign key
   referencing the table describing the **contained resource (or
   subcollection)**._
   
   _Please see the relevant sections of the_
   [Level 1 JSON Schema](../draft-C2M2_JSON_Schema_datapackage_specs/C2M2_Level_1.datapackage.json)
   _to find all table-specific field names and foreign-key constraints._

##### Association tables: inter-entity linkages

   * `file_describes_subject`
   * `file_describes_biosample`
   * `biosample_from_subject`
   * `collection_defined_by_project`
   
   _As with the containment association tables, records in these tables
   will contain four fields, encoding two foreign keys: one (composite
   `id_namespace+local_id`) key per entity involved in the particular
   relationship being asserted by each record._
   
   _Table names define relationship types, and are (with the exception
   of_ `collection_defined_by_project`_) somewhat nonspecific by design.
   Note in particular that relationships between core entities represented
   here **may** mask transitively-collapsed versions of more complex
   relationship networks in the native DCC metadataset. The specification
   of precise rules governing native-to-C2M2 metadata mappings (or
   approximations) are left to DCC serialization staff and relevant
   investigators; CFDE staff will be available as needed to offer
   feedback and guidance when navigating these issues._
   
   _Please see the relevant sections of the_
   [Level 1 JSON Schema](../draft-C2M2_JSON_Schema_datapackage_specs/C2M2_Level_1.datapackage.json)
   _to find all table-specific field names and foreign-key constraints._

##### Association table: taxonomy and the `subject` entity: the `subject_role_taxonomy` table

   _The_ `subject_role_taxonomy` _ "categorical association" table enables
   the attachment of taxonomic labels (NCBI Taxonomy Database identifiers, of the form_
   `/^NCBI:txid[0-9]+$/` _and stored for reference locally in the C2M2_
   `ncbi_taxonomy` _table) to C2M2_ `subject` _entities in a variety of
   ways, depending on_ `subject_granularity`_, using_ `subject_role` _values to specify
   the qualifying semantic or ontological context that should be applied to
   each taxonomic label._

   * `subject_granularity`_:_ `subject` _multiplicity specifier:_
      * _for each_ `subject` _record, pick one of
   [these values](../draft-C2M2_internal_CFDE_CV_tables/subject_granularity.tsv)
   and include its_ `id` _in the_ `granularity` _field in that record._
   * `subject_role`: _constituent relationship to intra-_`subject` _system:_
      * _each_ `subject_granularity` _corresponds to a subset of
      [these values](../draft-C2M2_internal_CFDE_CV_tables/subject_role.tsv),
      each of which can be labeled independently with NCBI Taxonomy Database
      IDs via_ `subject_role_taxonomy`.
   * `subject_role_taxonomy`: _Putting it all together: this association table
   stores **three items per record**, connecting components of_ `subject` _entities
   (_`subject_role`_s) to taxonomic assignments:_
      * _A (binary:_ `{ subject.id_namespace, subject.local_id }`_) key identifying a C2M2_ `subject` _entity record_
      * _An enumerated category code (the `id` field in [this table](../draft-C2M2_internal_CFDE_CV_tables/subject_role.tsv)) denoting a_ `subject_role` _contextual qualifier_
      * _A (unitary:_ `{ ncbi_taxonomy.id }`_) ID denoting an NCBI Taxonomy Database
      entry classifying the given_ `subject` _by way of the given_ `subject_role`

   _Please refer to the definition of_ `subject_role_taxonomy` _in the_
   [Level 1 JSON Schema](../draft-C2M2_JSON_Schema_datapackage_specs/C2M2_Level_1.datapackage.json)
   _to find all technical details (field names and foreign-key constraints)._

##### Controlled vocabularies and term tables

   * _CVs in use at Level 1:_
      * _**assay\_type (OBI):** used to describe **types of experiment** that can be produce
      1_ `file`_s_
      * _**anatomy (Uberon):** used to specify the **physiological source location**
      in or on the_ `subject` _from which a_ `biosample` _was derived_
      * _**data\_type (EDAM):** used to categorize the abstract **information content** of
      a_ `file` _(e.g. "this is sequence data")_
      * _**file\_format (EDAM):** used to denote the **digital format or encoding** of
      a_ `file` _(e.g. "this is a FASTQ file")_
      * _**ncbi\_taxonomy (...NCBI Taxonomy :):** used to **link**_ **`subject`** _**entity
      records to taxonomic labels** (cf above, §"Association table: taxonomy and the_ `subject`
      _entity: the_ `subject_role_taxonomy` _table")_
   * _**general guidance on usage:**_
      * _store bare CV terms (conforming to the pattern
      constraints specified for each CV's term set in the ER diagram, above) in the
      relevant entity-table fields (represented in the diagram as the sources of
      dotted green arrows)_
      * _for the moment -- with respect to deciding how to select
      terms -- just do the best you can by picking through the term sets provided by the
      given CVs_
      * _feel free to use more general ancestor terms if sufficiently-specific
      terms aren't available in a particular ontological CV_
      * _aggressively leave blank CV-field values for any records that wind up causing
      you the slightest bit of trouble._
      * `#c2m2-internal-note: See the wish list (two bullets below) for initial notes on improving the
      engineering solutions for this topic after the demo. Too many important issues remain
      to be studied, argued, decided, implemented and tested for us to make CV management
      more than a quick and intellectually unsatisfying kludge until the next round of
      active model development. (Note that now that we've actually built a few C2M2 metadata instance
      collections, we can begin systematically comparing notes to help us all get a much better collective
      grip on some of the problems that will need resolution in this area, and we're going to want to
      hear from DCCs which haven't yet submitted metadata as well, so they can help us anticipate difficulties.)`
   * [CV term scanner script](../draft-C2M2_external_CV_term_table_generator_script/build_term_tables.py):
   	_auto-builds (green) CV term tables_
      * _executed during BDBag-preparation stage, after core TSVs have been built_
      * _inflates (bare) CV terms cited in core-entity table fields into corresponding CV
      term-usage tables_
      * _auto-loads and populates display-layer term-decorator data (name,
      description) from relevant (versioned) CV reference files_
      * _usage:_
         * _change "_`USER-DEFINED PARAMETERS`_" section to match your local directory
      configuration_
         * _make sure the prerequisite files are in the right directories_
         * _then just run the script without arguments_
   * **non-optional wish list:** Everything listed in this segment must be
   carefully addressed and drafted in order to produce a mature policy on controlled vocabulary usage.
      * _explicit **version control** policy for reference CVs_
      * _detailed plan for handling app-layer aggregations of CV-term query
      results to **best serve users' search requests**:_
         * _**LCA computation** and implicit matching of terms via shared ontological lineage_
         * _keyword-set association/**tagging**/decoration_
         * _**synonym handling**_
         * _etc._
      * _policy specifying (or standardizing or prohibiting or ...?) a **term-addition
      request process** between CFDE and CV owners (active and ongoing between HMP
      and OBI, e.g.: terms are being added on request; CV managers are responsive),
      driven by usage needs identified by DCC clients_
      * _are **URIs better than bare CV terms** in terms of C2M2 field values?_
         * _what sort of URI support do CVs already provide?_
         * _how deeply can we leverage their own preexisting constructs without
         having to handle maintenance, synchrony, version, etc., issues ourselves?_
         * _can we establish a uniform URI policy to cover all C2M2-referenced CVs,
         or will we need to establish multiple policies for different CVs?_
      * _establish and execute some sort of survey process to create
      consensus on **which particular CVs look like the best final selections**
      to serve as sanctioned C2M2 reference sets (e.g. OBI vs. BAO); criteria:_
      	* _how comprehensive is a CV's coverage of the relevant ontological space?_
      	* _how responsive are the CV owners to change requests?_
      * _detailed ETL-construction usage plan: should we pre-select sub-vocabularies of
      sanctioned CVs to distribute to ETL generators, updating these CFDE-blessed
      CV subsets on an ongoing basis (as new term requirements roll in from
      client metadata sources (DCCs) as they try to model their respective datasets)?_

#### Level 1 metadata submission examples: Data Package JSON Schema and example TSVs

A JSON Schema document -- implementing
[Frictionless Data](https://frictionlessdata.io/)'s
"[Data Package](https://frictionlessdata.io/data-package/)"
container meta-specification -- defining the Level 1 TSV collection is
[here](../draft-C2M2_JSON_Schema_datapackage_specs/C2M2_Level_1.datapackage.json);
an example Level-1-compliant TSV submission collection can be found
[here](https://osf.io/uadny/) for inspection in two alternative forms: (1) a bare
collection of TSV files, and (2) a single packaged BDBag archive file containing those
TSVs along with some packaging/manifest metadata. (DCCs will package each C2M2
submission as one of these BDBags: we provide a valid one here for reference.)

--------------------------------------------------------------------------------

### Level 2

_C2M2 Level 2 is currently being drafted: publication of a complete specification is expected by the end of 2020._

1. _**New modeling concept checklist:**_
	* clinical **visit data**
	* modular **experimental flow (`protocol`)**
	* resource (entity) **provenance (`[data|material]_event` network)**
	* structured addressbook for documenting and linking **organizations
	(`common_fund_program`), roles/personae and actual people** to C2M2 metadata
	* **protected data**
	* full elaboration of scientific attributes of C2M2 entities using
	**controlled-vocabulary metadata decorations**
	   * i.e., substrate data for facet-search targets
	   * e.g., Level 1's `anatomy`, `assay_type`, `ncbi_taxonomy`, etc.)
	   * [enumerate requirements and scope for more complex modeling of scientific
	   metadata_]

--------------------------------------------------------------------------------
